pkg/products/marin3r/rateLimitService.go

```diff
@@ -0,0 +1,322 @@
+package marin3r
+
+import (
+	"context"
+	"fmt"
+
+	integreatlyv1alpha1 "github.com/integr8ly/integreatly-operator/pkg/apis/integreatly/v1alpha1"
+	"gopkg.in/yaml.v2"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type RateLimitServiceReconciler struct {
+	Namespace       string
+	RedisSecretName string
+	StatsdConfig    *StatsdConfig
+}
+
+type StatsdConfig struct {
+	Host string
+	Port string
+}
+
+func NewRateLimitServiceReconciler(namespace, redisSecretName string) *RateLimitServiceReconciler {
+	return &RateLimitServiceReconciler{
+		Namespace:       namespace,
+		RedisSecretName: redisSecretName,
+	}
+}
+
+// Config types, taken from unexported types in ratelimit service:
+// https://github.com/envoyproxy/ratelimit/blob/master/src/config/config_impl.go#L15-L44
+
+type yamlRateLimit struct {
+	RequestsPerUnit uint32 `yaml:"requests_per_unit"`
+	Unit            string
+}
+
+type yamlDescriptor struct {
+	Key         string
+	Value       string
+	RateLimit   *yamlRateLimit `yaml:"rate_limit"`
+	Descriptors []yamlDescriptor
+}
+
+type yamlRoot struct {
+	Domain      string
+	Descriptors []yamlDescriptor
+}
+
+// ReconcileRateLimitService creates the resources to deploy the rate limit service
+// It reconciles a ConfigMap to configure the service, a Deployment to run it, and
+// exposes it as a Service
+func (r *RateLimitServiceReconciler) ReconcileRateLimitService(ctx context.Context, client k8sclient.Client) (integreatlyv1alpha1.StatusPhase, error) {
+	phase, err := r.reconcileConfigMap(ctx, client)
+	if err != nil {
+		return integreatlyv1alpha1.PhaseFailed, err
+	}
+
+	if phase == integreatlyv1alpha1.PhaseInProgress {
+		return phase, nil
+	}
```

Comment:
Is this needed as `r.reconcileConfigMap` doesn't look to return a `PhaseInProgress` ? :thinking:

Reply:
It is not. I wrote it in advance thinking that it would be needed but forgot to remove it. I'll get rid of that block :+1: